src/QsCompiler/CompilerOptimizations/ConstantPropagation.fs

```diff
@@ -71,8 +60,68 @@ type ConstantPropagator(compiledCallables: ImmutableDictionary<QsQualifiedName,
                     if isLiteral rhs.Expression callableDict then
                         fillVars vars (StringTuple.fromSymbolTuple lhs, rhs.Expression)
-                        declarations <- declarations @ [sprintf "%O = %O" (StringTuple.fromSymbolTuple lhs) (prettyPrint rhs.Expression)]
                         // printfn "Found constant declaration: %O = %O" (StringTuple.fromSymbolTuple lhs) (prettyPrint rhs.Expression)
                 QsBinding<TypedExpression>.New stm.Kind (lhs, rhs) |> QsVariableDeclaration
+
+            override statementKind.onConditionalStatement stm = 
+                let cbList, cbListEnd =
+                    stm.ConditionalBlocks |> Seq.fold (fun s (c, b) ->
+                        let cond, block = statementKind.onPositionedBlock (Some c, b)
+                        match cond.Value.Expression with
+                        | BoolLiteral true -> s @ [None, block]
+                        | BoolLiteral false -> s
+                        | _ -> s @ [cond, block]
+                    ) [] |> List.ofSeq |> takeWhilePlus1 (fun (c, b) -> c <> None)
+                let defaultCase = stm.Default |> QsNullable<_>.Map (fun b -> statementKind.onPositionedBlock (None, b) |> snd)
+
+                let newDefault =
+                    match cbListEnd, defaultCase with
+                    | Some (_, a), _ -> Value a
+                    | _, Value a -> Value a
+                    | _ -> Null
+                match cbList, newDefault with
+                | [], Value x ->
+                    x.Body |> QsScopeStatement.New |> QsScopeStatement
+                | [], Null ->
+                    QsScope.New ([], LocalDeclarations.New []) |> QsScopeStatement.New |> QsScopeStatement
+                | _ ->
+                    let cases = cbList |> Seq.map (fun (c, b) -> (Option.get c, b))
+                    QsConditionalStatement.New (cases, defaultCase) |> QsConditionalStatement
+
+            override statementKind.onForStatement stm =
+                let loopVar = fst stm.LoopItem |> statementKind.onSymbolTuple
+                let iterVals = statementKind.ExpressionTransformation stm.IterationValues
+                let iterValsAsSeq =
+                    match iterVals.Expression with
+                    | RangeLiteral _ -> 
+                        rangeLiteralToSeq iterVals.Expression |> Seq.map (fun x -> wrapExpr (IntLiteral x) Int) |> Some
+                    | ValueArray va -> va :> seq<_> |> Some
+                    | _ -> None
+
+                match iterValsAsSeq with
+                | Some s ->
+                    let iterRange = 
+                        s |> Seq.map (fun x ->
+                            let variableDecl = QsBinding.New ImmutableBinding (loopVar, x) |> QsVariableDeclaration
+                            let variableDeclStatement = {
+                                Statement = variableDecl
+                                SymbolDeclarations = stm.Body.KnownSymbols
```

Comment:
… and with the PR that handles the variable naming you need to make sure to adapt the known symbols accordingly, and also the declarations attached to each scope.